src/server/controllers/changePassword.ts

```diff
@@ -0,0 +1,131 @@
+  const errors: Array<FieldError> = [];
+
+  if (!password) {
+    errors.push({
+      field: 'password',
+      message: ChangePasswordErrors.PASSWORD_BLANK,
+    });
+  } else if (password.length < 8 || password.length > 72) {
+    errors.push({
+      field: 'password',
+      message: ChangePasswordErrors.PASSWORD_LENGTH,
+    });
+  }
+
+  return errors;
+};
+
+export const checkResetPasswordTokenController = (
+  setPasswordPagePath: string,
+  setPasswordPageTitle: string,
+  resendEmailPagePath: string,
+  resendEmailPageTitle: string,
+) =>
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    let state = res.locals;
+    const { token } = req.params;
+
+    try {
+      state = deepmerge(state, {
+        pageData: {
+          browserName: getBrowserNameFromUserAgent(req.header('User-Agent')),
+          email: await validateToken(token, req.ip),
+        },
+      });
+
+      const html = renderer(`${setPasswordPagePath}/${token}`, {
+        requestState: state,
+        pageTitle: setPasswordPageTitle,
+      });
+      return res.type('html').send(html);
+    } catch (error) {
+      logger.error(`${req.method} ${req.originalUrl}  Error`, error);
+
+      return res.type('html').send(
+        renderer(`${resendEmailPagePath}${Routes.RESEND}`, {
+          requestState: state,
+          pageTitle: resendEmailPageTitle,
+        }),
+      );
+    }
+  });
+
+export const setPasswordTokenController = (
+  setPasswordPath: string,
+  setPasswordPageTitle: string,
+  successCallback: (res: ResponseWithRequestState) => void,
+) =>
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    let state = res.locals;
+
+    const { token } = req.params;
+    const { password } = req.body;
+
+    state = deepmerge(state, {
+      pageData: {
+        browserName: getBrowserNameFromUserAgent(req.header('User-Agent')),
+      },
+    });
+
+    try {
+      const fieldErrors = validatePasswordField(password);
+
+      if (fieldErrors.length) {
+        state = deepmerge(state, {
+          pageData: {
+            email: await validateToken(token, req.ip),
+            fieldErrors,
+          },
+        });
+        const html = renderer(`${setPasswordPath}/${token}`, {
+          requestState: state,
+          pageTitle: setPasswordPageTitle,
+        });
+        return res.status(422).type('html').send(html);
+      }
+
+      const cookies = await changePassword(password, token, req.ip);
+
+      setIDAPICookies(res, cookies);
+
+      trackMetric(Metrics.ACCOUNT_VERIFICATION_SUCCESS);
+      trackMetric(Metrics.CHANGE_PASSWORD_SUCCESS);
+
+      return successCallback(res);
+    } catch (error) {
+      const { message, status } = error;
+      logger.error(`${req.method} ${req.originalUrl}  Error`, error);
+
+      trackMetric(Metrics.ACCOUNT_VERIFICATION_FAILURE);
+      trackMetric(Metrics.CHANGE_PASSWORD_FAILURE);
```

Comment:
It feels a bit iffy to have two success and failure metrics for one event. I get that there are two things happening here, but still, writing two events to the log might cause confusion later on?

Reply:
Is this perhaps a naming problem instead?

Reply:
Yeah, it's because the change password endpoint in IDAPI is now doing two things, a) ACCOUNT_VERIFICATION and b) CHANGE_PASSWORD which we'd want to track separately.
I think I'll add a comment to the code to explain why there's two things going in.